Feature/refactor merge to graph

aaronchongth opened the pull request:
* migrated `FleetAdapterNode::compute_plan_starts` to `rmf_traffic::agv::Planner::compute_plan_starts`

* removed undesired behaviour of merging into closest waypoint or lane returns `rmf_utils::nullopt` whenever it is outside of desired distance away from all waypoints or lanes

* parameterized merging distance for waypoints or lanes

* added tests for desired behaviours of `rmf_traffic::agv::Planner::computer_plan_starts`

TODO:

* might consider refactor distance finding math into the `Waypoint` or `Lane` classes

mxgrey commented on rmf_traffic/test/unit/agv/test_Planner.cpp:
```diff
@@ -2342,3 +2342,588 @@ SCENARIO("Test starts using graph with non-colinear waypoints")
+      Planner::StartSet start_set = start_set_opt.value();
+      CHECK(!start_set.empty());
+      CHECK(start_set.size() == 1);
+      CHECK(start_set[0].waypoint() == 0);
```
We should probably add a check for which lane was chosen here, as well as in the other tests which expect a lane to specified. For tests which do not expect a lane to specified, we should check that no lane was specified, i.e. `CHECK(!start_set[0].lane());`.

aaronchongth replied:
ah, didn't know that that was something that could be checked, cool

mxgrey commented on rmf_traffic/test/unit/agv/test_Planner.cpp:
```diff
@@ -2342,3 +2342,588 @@ SCENARIO("Test starts using graph with non-colinear waypoints")
+    graph.add_waypoint(test_map_name, {10.0, 0}); // 2
+    REQUIRE(graph.num_waypoints() == 3);
+
+    graph.add_lane(0, 1); // 0
```
What if we add a lane from 1 -> 0 (`graph.add_lane(1, 0);`)? Every lane that is added is unidirectional, so it would be good to test what happens when there are bidirectional connections between waypoints.

aaronchongth replied:
gotcha, should have checked what the other tests were doing too :boom:

aaronchongth commented on rmf_traffic/test/unit/agv/test_Planner.cpp:
```diff
@@ -2342,3 +2342,588 @@ SCENARIO("Test starts using graph with non-colinear waypoints")
+      rmf_utils::optional<Planner::StartSet> start_set_opt =
+          planner.compute_plan_starts(robot_loc, initial_time, 0.1, 1.0);
+      CHECK(start_set_opt);
+      Planner::StartSet start_set = start_set_opt.value();
```
ooohh, makes sense, will most likely start using `const auto&` more

aaronchongth commented on rmf_traffic/include/rmf_traffic/agv/Planner.hpp:
```diff
@@ -333,6 +333,37 @@ class Planner
+  /// Produces a set of starting conditions in order to start planning. This
+  /// method attempts to find the most suitable starting nodes within the graph
+  /// for merging, planning and execution of plans. If none of the waypoints in
+  /// the graph fulfils the requirements, a nullopt will be returned.
```
ah, it was probably a decision I made before that conversation which we decided to forgo that take-closest-lane-anyway behaviour. Thanks for catching that

aaronchongth commented on rmf_traffic/include/rmf_traffic/agv/Planner.hpp:
```diff
@@ -333,6 +333,37 @@ class Planner
+  ///   The maximum distance allowed to automatically merge onto a lane, i.e.
+  ///   adding the lane's entry and exit waypoints as potential starts. default
+  ///   value as 1.0 meters.
+  rmf_utils::optional<StartSet> compute_plan_starts(
```
Would you be open to the idea of having it as something under a new header called `Behaviour`? Probably under `Behaviour::PlanStarter` with `Behaviour::PlanStarter::Options`.

This might be helpful when we start to identify other behaviours we might want AGVs to follow within RMF.

aaronchongth replied:
just a note about our conversation, `compute_plan_starts` will be migrated to a utility function under `Planner.hpp`, with all the needed parameters passed in

aaronchongth commented:
* refactored to make it a utility function

* corrected some mistakes, after understanding more of what `Planner` expects

* plenty more tests written

mxgrey commented on rmf_traffic/include/rmf_traffic/agv/Planner.hpp:
```diff
@@ -573,6 +544,47 @@ class Plan
+///
+/// \param[in] start_time
+///   The starting time that will be attributed to all the generated starts
+///   to compute a new plan. In some occations, users will want to add small
```
```suggestion
///   to compute a new plan. In some occasions, users will want to add small
```

mxgrey commented on rmf_traffic/src/rmf_traffic/agv/Planner.cpp:
```diff
@@ -738,5 +602,105 @@ const Planner::Configuration& Plan::get_configuration() const
+      }
+    }
+    // If it's between the entry and the exit waypoints, then we should
+    // compute it's distance away from the lane line.
```
```suggestion
    // compute its distance away from the lane line.
```

mxgrey commented on rmf_traffic/test/unit/agv/test_Planner.cpp:
```diff
@@ -2855,277 +2856,628 @@ SCENARIO("Test computing Starts from coordinates")
-      
       CHECK(start_set[i].location());
+      CHECK(start_set[i].lane());
+      std::size_t& lane_index = start_set[i].lane().value();
```
Nitpick: This should be either `std::size_t` or `const std::size_t&` if you don't intend to modify the value.

mxgrey commented on rmf_traffic/test/unit/agv/test_Planner.cpp:
```diff
@@ -2855,277 +2856,628 @@ SCENARIO("Test computing Starts from coordinates")
+    graph.add_lane(1, 2); // 1
+    REQUIRE(graph.num_lanes() == 2);
+
+    planner = Planner{Planner::Configuration{graph, traits}, default_options};
```
It looks to me like this `Planner` instance only ever gets used to store the `Graph`, so it's not really accomplishing anything. It may be worth checking the other start tests to see if they are similarly using an unnecessary `Planner` instance.

aaronchongth replied:
thanks for catching this!

mxgrey commented on rmf_traffic/test/unit/agv/test_Planner.cpp:
```diff
@@ -2855,277 +2856,628 @@ SCENARIO("Test computing Starts from coordinates")
+      CHECK((wp_index == 1 || wp_index == 2 || wp_index == 3));
+      CHECK(start_set[i].location());
+      CHECK(start_set[i].lane());
+      std::size_t& lane_index = start_set[i].lane().value();
```
Again use `std::size_t` or `const std::size_t&`
